### ISSUE TITLE
network loses 2 invokes, after some restarts, after PEER0 wanders off

### ISSUE THREAD

**Issue by scottz64** [hyperledger-archives/fabric#2148]
## Description

Two Invoke transactions appear to be lost at one point, during a sequence of stopping and restarting peer nodes (in my docker network of 4 peers running in batch mode with security). Network nodes to eventually resume consensus but they apparently lost 2 queued invoke transactions at some point.
## Describe How to Reproduce

See detailed logs attached which describe test scenario, both passing and failing examples of running the same test. I used our GO SDK testcase name: 2loopBounceDoubleEvery5th.go
This same testcase was run several times. Sometimes it passed just fine, and other times an error occurred once (or more than once in some cases). The basic idea for the test steps is as follows. Between each step, I send an invoke to each peer that is running, and I get acknowledgement that it is received. I sleep/delay between each step. In the following sequence, all 4 nodes are in sync with B=1000052 before these steps. After this starting point, peer0 never seems to rejoin the consensus and syncs with others until B=1000104 (at which point B should equal 1000106). You can see many more details and the logfiles, in the attachments.

STOP PEERS():  PEER0
START PEERS():  PEER0
STOP PEERS():  PEER0  PEER1
START PEERS():  PEER0  PEER1
STOP PEERS():  PEER2
START PEERS():  PEER2
STOP PEERS():  PEER3
START PEERS():  PEER3
STOP PEERS():  PEER0
START PEERS():  PEER0
STOP PEERS():  PEER1
START PEERS():  PEER1
STOP PEERS():  PEER2  PEER3
START PEERS():  PEER2  PEER3
QUERY RESULTS: 
FAIL on PEER0: EXPECTED/ACTUAL: A=999894/999896, B=1000106/1000104. ****_FAIL**_**
FAIL on PEER1: EXPECTED/ACTUAL: A=999894/999900, B=1000106/1000100. ****_FAIL**_**
FAIL on PEER2: EXPECTED/ACTUAL: A=999894/999896, B=1000106/1000104. ****_FAIL**_**
FAIL on PEER3: EXPECTED/ACTUAL: A=999894/999896, B=1000106/1000104. ****_FAIL**_**

[Lose2of3InvokesAfterSomeRestarts.tar.gz](https://github.com/hyperledger/fabric/files/355143/Lose2of3InvokesAfterSomeRestarts.tar.gz)

[pass2loop_for_comparison.tar.gz](https://github.com/hyperledger/fabric/files/355144/pass2loop_for_comparison.tar.gz)

**Comment by corecode** [comment]
This is quite expected, because there is no complete chain of custody.  This is something we can look at for consensus v2.

**Comment by scottz64** [comment]
@corecode @tuand27613 @kchristidis @nickgaski  If I understand you correctly, then I would expect it to happen every time when I run the same testcase, but it does not. Can you identify which transactions were lost, and the key steps in the test sequence that led to that?

There may be something that I am missing; here is my understanding and expectations:
- Sending an Invoke to a running Peer, while there are not enough peers in the network for consensus, causes it to get queued; then it gets processed when enough peers join the network. 
- I do not send any Invokes to a peer that is not running (and thus I do not miscount). 
- After having sent an Invoke to a running peer (when only 2 peers are running), I do not stop/restart that peer until after others are restarted and consensus is resumed.

I suspect there is a real problem here around peer 0, which didn't seem to fully rejoin consensus network (and catchup) in the middle steps even when peer1 was stopped.

**Comment by corecode** [comment]
I'm just saying that knowing the code, there are several places where this could happen.

**Comment by yeasy** [comment]
Even with all nodes running normally, the transactions can lose with sequentially requests.

**Comment by scottz64** [comment]
@corecode I agree that a complete chain of custody is desirable and hopefully can be addressed in v2.

But if it is a chain of custody problem, then why doesn't it happen every time I run the test sequence? I still wonder why sometimes peer0 does not recover, during the middle steps of the test.

@yeasy My testcase just sends one invoke transaction to each running VP, between each step (stop or restart), so there is no concern about overloading the peers. Under what conditions could transactions be lost somehow with all peers running?

**Comment by yeasy** [comment]
IMHO, the losing result itself is not serious problem. But mostly, the return result just looks normally, which can confuse people.
Maybe we should add more return code to differentiate the cases.

**Comment by corecode** [comment]
@yeasy return code for what?  if we lose a transaction, we don't know that we lost it :)

**Comment by yeasy** [comment]
@corecode I mean, when we invoke some transactions, can the system return some simple code to indicate whether the invocation succeeds or not?
